File: internal/grpc/typefacts.go
package grpc

import (
	"sync"

	pb "github.com/Open-MBEE/OpenSysML/api/proto"
	"github.com/Open-MBEE/OpenSysML/internal/core/passes"
	"github.com/Open-MBEE/OpenSysML/internal/protoconv"
	"github.com/Open-MBEE/OpenSysML/internal/semantic/resolve"
	"github.com/Open-MBEE/OpenSysML/internal/semantic/semantics"
	"github.com/Open-MBEE/OpenSysML/internal/semantic/symbols"
	"github.com/Open-MBEE/OpenSysML/internal/syntax/ast"
)

// fqnScalarQuantityValue is the library type every quantity value specializes.
const fqnScalarQuantityValue = "Quantities::ScalarQuantityValue"

// SymbolContext holds what converting a symbol to proto needs besides the
// symbol: its index, and the resolver and semantic model that answer what its
// declared names refer to. Their memoization is why it is shared per model.
//
// The resolver and semantic model memoize into plain maps, so a shared context
// must be used by one goroutine at a time; Lock serializes concurrent
// conversions of the same model.
type SymbolContext struct {
	mu sync.Mutex

	Index     *symbols.Index
	Resolver  *resolve.Resolver
	Semantics *semantics.Model
}

// Lock takes exclusive use of the context's resolver and semantic model, and
// returns the function releasing it.
func (sc *SymbolContext) Lock() func() {
	sc.mu.Lock()
	// Resolution during conversion is speculative: a name that does not resolve
	// is reported as unresolved type facts, not as a model diagnostic. Dropping
	// what it appended keeps the shared resolver's diagnostics from growing with
	// every request.
	diags := len(sc.Resolver.Diagnostics)
	return func() {
		if len(sc.Resolver.Diagnostics) > diags {
			sc.Resolver.Diagnostics = sc.Resolver.Diagnostics[:diags]
		}
		sc.mu.Unlock()
	}
}

// NewSymbolContext builds a conversion context over a symbol index.
func NewSymbolContext(idx *symbols.Index) *SymbolContext {
	resolver := resolve.New(idx)
	sem := passes.NewTypedModel(resolver)
	resolver.SetModel(sem)
	return &SymbolContext{Index: idx, Resolver: resolver, Semantics: sem}
}

// relationshipKindName maps a generalization relationship to the name the wire
// reports it under. A relationship that is not a generalization has none.
func relationshipKindName(k ast.RelationshipKind) string {
	switch k {
	case ast.RelSpecializes:
		return "specializes"
	case ast.RelSubsets:
		return "subsets"
	case ast.RelRedefines:
		return "redefines"
	case ast.RelTyping:
		return "typing"
	default:
		return ""
	}
}

// relationshipName returns the qualified name a relationship targets,
// unwrapping the feature reference a usage's typing may be written as.
func relationshipName(rel *ast.Relationship) *ast.QualifiedName {
	if rel == nil || rel.Target == nil {
		return nil
	}
	target := rel.Target
	if fr, ok := target.(*ast.FeatureReference); ok {
		target = fr.Name
	}
	qn, ok := target.(*ast.QualifiedName)
	if !ok {
		return nil
	}
	return qn
}

// specializationsOf reports every generalization edge a symbol declares, in
// declaration order, with each target resolved.
func (sc *SymbolContext) specializationsOf(sym *symbols.Symbol) []*pb.Specialization {
	var out []*pb.Specialization
	for _, rel := range semantics.RelationshipsOf(sym) {
		kind := relationshipKindName(rel.Kind)
		if kind == "" {
			continue
		}
		qn := relationshipName(rel)
		if qn == nil {
			continue
		}
		spec := &pb.Specialization{
			Kind:     kind,
			Declared: qn.Text(),
		}
		if target := sc.resolveFrom(sym, qn); target != nil {
			spec.TargetId = sc.Index.GetFQN(target)
			spec.TargetKind = target.Kind.String()
		}
		out = append(out, spec)
	}
	return out
}

// resolveFrom resolves a qualified name written in sym's declaration,
// following an alias to what it names.
func (sc *SymbolContext) resolveFrom(sym *symbols.Symbol, qn *ast.QualifiedName) *symbols.Symbol {
	target, ok := sc.Resolver.ResolveQualified(sym.OwnerScope, qn)
	if !ok || target == nil {
		return nil
	}
	if alias, ok := sc.Resolver.ResolveAliasTarget(target); ok {
		target = alias
	}
	return target
}

// multiplicityOf reports the multiplicity a symbol declares, or nil when it
// declares none. An unevaluable bound is reported empty rather than guessed at.
func (sc *SymbolContext) multiplicityOf(sym *symbols.Symbol) *pb.MultiplicityInfo {
	rng, ok := sc.Semantics.MultiplicityOf(sym)
	if !ok {
		return nil
	}
	return &pb.MultiplicityInfo{
		Lower: protoconv.BoundText(rng.Lower),
		Upper: protoconv.BoundText(rng.Upper),
	}
}

// typeInfoOf derives the static type facts of a def or usage. Anything the
// model cannot derive is left empty rather than guessed at.
func (sc *SymbolContext) typeInfoOf(sym *symbols.Symbol) *pb.TypeInfo {
	switch decl := sym.Decl.(type) {
	case *ast.Usage:
		return sc.usageTypeInfo(sym, decl)
	case *ast.Definition:
		return sc.definitionTypeInfo(sym)
	default:
		return nil
	}
}

// usageTypeInfo derives the type facts of a usage. An untyped usage still has a
// primitive when its default value determines one, reported as inferred.
func (sc *SymbolContext) usageTypeInfo(sym *symbols.Symbol, decl *ast.Usage) *pb.TypeInfo {
	info := &pb.TypeInfo{}

	// Declared type: the first typing relationship, which is the usage's own
	// type. Subsetting and redefinition are reported as specializations.
	for _, rel := range decl.Relationships {
		if rel.Kind != ast.RelTyping {
			continue
		}
		qn := relationshipName(rel)
		if qn == nil {
			continue
		}
		info.Declared = qn.Text()
		if target := sc.resolveFrom(sym, qn); target != nil {
			info.ResolvedId = sc.Index.GetFQN(target)
			info.ResolvedKind = target.Kind.String()
		}
		break
	}

	// PrimTypeOf walks the generalization graph, so it covers a type inherited
	// through subsetting or redefinition as well as a declared one.
	if prim := sc.Semantics.PrimTypeOf(sym); prim != semantics.PrimUnknown {
		info.Primitive = prim.String()
		info.PrimitiveSource = "declared"
	} else if prim, ok := primOfValue(sc.Semantics, decl.Value); ok {
		info.Primitive = prim
		info.PrimitiveSource = "value"
	}

	sc.markQuantity(sym, info, decl.Value)
	return info
}

// definitionTypeInfo derives the type facts of a definition: the library scalar
// it classifies as, and whether it is a quantity value type.
func (sc *SymbolContext) definitionTypeInfo(sym *symbols.Symbol) *pb.TypeInfo {
	info := &pb.TypeInfo{}
	if prim := sc.Semantics.PrimTypeOf(sym); prim != semantics.PrimUnknown {
		info.Primitive = prim.String()
		info.PrimitiveSource = "declared"
	}
	sc.markQuantity(sym, info, nil)
	return info
}

// markQuantity records that an element's values carry a measurement unit: its
// type is a quantity value type, or its default value names a unit.
func (sc *SymbolContext) markQuantity(sym *symbols.Symbol, info *pb.TypeInfo, value ast.Node) {
	if idx, ok := value.(*ast.IndexExpr); ok && idx.Bracket {
		if _, err := sc.Semantics.UnitTermOfExpr(sym.OwnerScope, idx.Index); err == nil {
			info.Quantity = true
			info.Unit = semantics.UnitExprText(idx.Index)
		}
	}
	if info.Quantity {
		return
	}
	quantityValue := sc.libSymbol(fqnScalarQuantityValue)
	if quantityValue == nil {
		return
	}
	// Typing is a generalization edge, so a usage conforms to its type's
	// supertypes: the same check covers a def and a usage.
	if sc.Semantics.Conforms(sym, quantityValue) {
		info.Quantity = true
	}
}

// libSymbol looks up a library element by qualified name, uniquely or not at
// all: an ambiguous name is no evidence of the library's element.
func (sc *SymbolContext) libSymbol(fqn string) *symbols.Symbol {
	matches := sc.Index.LookupQualified(fqn)
	if len(matches) != 1 {
		return nil
	}
	return matches[0]
}

// primOfValue classifies a default value expression. Only a constant-evaluable
// value or a string literal determines a type; anything else stays underivable.
func primOfValue(sem *semantics.Model, value ast.Node) (string, bool) {
	if value == nil {
		return "", false
	}
	if _, isStr := value.(*ast.LiteralString); isStr {
		return semantics.PrimString.String(), true
	}
	val, ok := sem.Eval(value)
	if !ok {
		return "", false
	}
	switch val.Kind {
	case semantics.ValInt:
		return semantics.PrimInteger.String(), true
	case semantics.ValReal:
		return semantics.PrimReal.String(), true
	case semantics.ValBool:
		return semantics.PrimBoolean.String(), true
	default:
		return "", false
	}
}

File: client/python/opensysml/typefacts.py
"""Static type facts a symbol carries: its type, multiplicity and supertypes.

These mirror the `TypeInfo`, `MultiplicityInfo` and `Specialization` messages the
service sends. They are plain frozen dataclasses so that code generation can be
exercised over synthetic symbol trees.
"""

from dataclasses import dataclass, field
from typing import Optional, Tuple

from opensysml.values import value_to_python


@dataclass(frozen=True)
class TypeFacts:
    """The static type of a usage, or the classification of a definition."""

    declared: str = ""
    """Type name as written; empty when none is declared."""

    resolved_id: str = ""
    """FQN of the resolved type; empty when unresolved or undeclared."""

    resolved_kind: str = ""
    """Symbol kind of the resolved type, e.g. ``partDef``."""

    primitive: str = ""
    """Library scalar the type reduces to, e.g. ``Real``; empty when it is not one."""

    primitive_source: str = ""
    """Origin of :attr:`primitive`: ``declared``, ``value`` or empty."""

    quantity: bool = False
    """Values carry a measurement unit."""

    unit: str = ""
    """Unit as written, when the default value names one."""

    @classmethod
    def from_pb(cls, pb_type_info) -> "TypeFacts":
        """Build from a ``TypeInfo`` protobuf message."""
        return cls(
            declared=pb_type_info.declared,
            resolved_id=pb_type_info.resolved_id,
            resolved_kind=pb_type_info.resolved_kind,
            primitive=pb_type_info.primitive,
            primitive_source=pb_type_info.primitive_source,
            quantity=pb_type_info.quantity,
            unit=pb_type_info.unit,
        )


@dataclass(frozen=True)
class Multiplicity:
    """A declared multiplicity range. A bound the service could not evaluate is empty."""

    lower: str = ""
    upper: str = ""

    @classmethod
    def from_pb(cls, pb_multiplicity) -> "Multiplicity":
        """Build from a ``MultiplicityInfo`` protobuf message."""
        return cls(lower=pb_multiplicity.lower, upper=pb_multiplicity.upper)

    @property
    def is_collection(self) -> Optional[bool]:
        """Whether the range admits more than one value, or None when unknown."""
        if self.upper == "":
            return None
        if self.upper == "*":
            return True
        try:
            return int(self.upper) > 1
        except ValueError:
            return None

    @property
    def is_optional(self) -> Optional[bool]:
        """Whether the range admits no value, or None when unknown."""
        if self.lower == "":
            return None
        try:
            return int(self.lower) == 0
        except ValueError:
            return None


@dataclass(frozen=True)
class Specialization:
    """One generalization edge: ``specializes``, ``subsets``, ``redefines`` or ``typing``."""

    kind: str
    declared: str = ""
    target_id: str = ""
    target_kind: str = ""

    @classmethod
    def from_pb(cls, pb_specialization) -> "Specialization":
        """Build from a ``Specialization`` protobuf message."""
        return cls(
            kind=pb_specialization.kind,
            declared=pb_specialization.declared,
            target_id=pb_specialization.target_id,
            target_kind=pb_specialization.target_kind,
        )


@dataclass(frozen=True)
class AttributeFacts:
    """One attribute an element has, own or inherited, as the service resolves it."""

    name: str
    type: str = ""
    """FQN of the resolved type, else the type as written, else the library scalar."""

    value: object = None
    """Default value, when it is a model-level constant; None when there is none."""

    unit: str = ""
    """Unit the default value is written in; empty when it carries none."""

    @classmethod
    def from_pb(cls, pb_attribute) -> "AttributeFacts":
        """Build from an ``AttributeInfo`` protobuf message."""
        value = None
        if pb_attribute.HasField("value"):
            value = value_to_python(pb_attribute.value)
        return cls(
            name=pb_attribute.name,
            type=pb_attribute.type,
            value=value,
            unit=pb_attribute.unit,
        )


@dataclass(frozen=True)
class SymbolFacts:
    """Everything code generation needs about one symbol."""

    id: str
    name: str
    kind: str
    type: Optional[TypeFacts] = None
    multiplicity: Optional[Multiplicity] = None
    specializations: Tuple[Specialization, ...] = field(default_factory=tuple)
    attributes: Tuple[AttributeFacts, ...] = field(default_factory=tuple)
    #: Library-inherited attributes ``attributes`` leaves out, counted rather
    #: than silently dropped.
    withheld_library_attributes: int = 0
